[MP Rewrite Part 5] Port previous planners to new framework

kylestach opened the pull request:
Most of these are not functional changes from the old planners, but
rather simple ports to the new primitives. The biggest changes are that
we now use `Replanner` static functions and manually manage our own
"previous trajectory" objects instead of having "behavior-level"
planners (i.e. settle) own instances of an RRT planner and interacting
with it through plan requests.

This is the final PR in the rewrite, and it should pass CI. At the point
where it passes CI and all PRs are ready for merge, each PR should be
merged into its parent in sequence.

kylestach committed:
[MP Rewrite Part 5] Port previous planners to new framework

Most of these are not functional changes from the old planners, but
rather simple ports to the new primitives. The biggest changes are that
we now use `Replanner` static functions and manually manage our own
"previous trajectory" objects instead of having "behavior-level"
planners (i.e. settle) own instances of an RRT planner and interacting
with it through plan requests.

This is the final PR in the rewrite, and it should pass CI. At the point
where it passes CI and all PRs are ready for merge, each PR should be
merged into its parent in sequence.

oswinso commented on soccer/planning/planner/CollectPlanner.hpp:
```diff
@@ -0,0 +1,156 @@
+/**
+ * @brief Planner that tries to move onto and gain control of a slow moving ball
+ *
+ * TODO: Clean up description
```
issue num

oswinso commented on soccer/planning/planner/LineKickPlanner.hpp:
```diff
@@ -0,0 +1,39 @@
+#include "planning/low_level/Replanner.hpp"
+
+namespace Planning {
+class LineKickPlanner : public PlannerForCommandType<LineKickCommand> {
```
docstring

oswinso reviewed (changes requested):
issue num & docstring

kylestach committed:
Merge branch 'mp-rewrite/primitives' into mp-rewrite/planner-ports

kylestach committed:
Address review comments

kylestach committed:
Merge branch 'mp-rewrite/primitives' into mp-rewrite/planner-ports

oswinso reviewed (approved):
lgtm

JNeiger commented on soccer/planning/planner/CollectPlanner.cpp:
```diff
@@ -0,0 +1,470 @@
+
+    // If we are close enough to the target point near the ball
+    // and almost the same speed we want, start slowing down
+    // TODO(Kyle): Check for ball sense?
```
Unrelated to this PR, but do we have angle information in these planners now? The old iteration only had XY information

kylestach replied:
Yeah, that's a big change to how these planners work. They now plan angles on the trajectory directly rather than returning an angle function.

JNeiger replied:
That was a big problem I had when trying to "match" the state machines inside the planner and in python. It was possible to have the ball behind us but close enough such that the planner thought we had the ball, but the python side knew we didn't. I don't think I ever found a good hack to make it completely consistent

kylestach replied:
I'm going to add a TODO issue for this. It should be much more reasonable to incorporate once we move to the "behavior planner" approach to these planners, where (in this case) it will actually have control and visibility over the dribbler.

kylestach replied:
#1518

JNeiger commented on soccer/planning/planner/CollectPlanner.cpp:
```diff
@@ -0,0 +1,470 @@
+
+    motionConstraints.maxAcceleration *= *_controlAccelScalePercent;
+    motionConstraints.maxSpeed =
+        std::min(currentSpeed, motionConstraints.maxSpeed);
```
This is the max speed thing I mentioned in #1513

JNeiger commented on soccer/planning/planner/LineKickPlanner.cpp:
```diff
@@ -0,0 +1,247 @@
+
+using namespace Geometry2d;
+Trajectory LineKickPlanner::plan(const PlanRequest& /* request */) {
+#if 0
```
LineKick is disabled?

kylestach replied:
I was going to leave it disabled in this PR, and make a separate PR into `mp-rewrite/base` where I actually port it over.

JNeiger commented on soccer/planning/planner/Planner.cpp:
```diff
@@ -0,0 +1,28 @@
+
+namespace Planning {
+
+#if 0
```
Disabled code?

kylestach replied:
Removed.

JNeiger commented on soccer/planning/planner/Planner.hpp:
```diff
@@ -0,0 +1,78 @@
+template <typename T>
+inline T applyLowPassFilter(const T& oldValue, const T& newValue, double gain) {
+    return gain * newValue + (1 - gain) * oldValue;
+}
```
Copy paste error?

kylestach replied:
Moving to utils in #1512

JNeiger commented on soccer/planning/planner/SettlePlanner.cpp:
```diff
@@ -0,0 +1,633 @@
+        //  Almost at end of the target path or
+        //  Already in line with the ball
+        //
+        // TODO(Kyle): Check ball sense?
```
Based on how I imagined dampen, it should already be in the dampen state before the ball hits. It should have some negative velocity when the ball hits and slow to a stop, minimizing the delta velocity between the robot and ball.

It was disabled through the config gains fairly early on during testing so no promises how well this actually works.

JNeiger commented on soccer/planning/planner/SettlePlanner.cpp:
```diff
@@ -0,0 +1,633 @@
+            ball.query_seconds_near(pathAtEnd, &nearPoint);
+        if (timeRemaining <
+                timeToIntersect + RJ::Seconds(*_interceptBufferTime) &&
+            nearPoint.distTo(pathAtEnd) < Replanner::goalPosChangeThreshold()) {
```
This will need to be tested, but you may run into problems with the `goalPosChangeThreshold` being larger than the `mouthWidth/2` so the reused path won't actually get you close enough to the ball path to actually catch it

JNeiger replied:
This is especially problematic since the "shortcut" stuff is cut out with this early return. The ball velocity is surprisingly noisy

kylestach replied:
Yeah, I'll remove this. It's no longer really an issue now that replanning is working a lot better.

kylestach committed:
Merge branch 'mp-rewrite/primitives' into mp-rewrite/planner-ports

kylestach committed:
Move applyLowPassFilter into common, add TODOs

kylestach committed:
Enable escape-obstacles planner

kylestach commented:
This now contains everything except for line kick, which I am planning on doing in a separate PR to mp-rewrite/base (logic being, it needs to share a lot of logic with settle and I'd like to clean that up a lot).

kylestach committed:
[MP Rewrite Part 6] Cleanup and fixes (#1512)

* Cleanup and fixes to make code run properly

* Fix test compilation and disable a trapezoidal motion profile edge case.

Some planner tests are still failing. This will be addressed in a later
commit.

* Prettify

* Make e2e planner tests pass.

Because they generate random scenarios, and the RRT planner is not
guaranteed to converge in a finite number of iterations (and the Bezier
approximation may be incorrect), these tests are currently flaky. That
should be fixed before this is merged to staging.

* Tidy fixes

* Linting

* Formatting

* Remove deleted file from CMakeLists, remove reuse condition from settle

* Don't replan in Collect::control

* Optimizations to geometry methods

* Reset and regenerate sim config

This deletes unnecessary variables and moves planning-related variables
into PathPlanner

* Formatting

* Add TODO in EscapeObstacles

* Tidy fixes

* Revert inconsistent clang-tidy recommended fix

* Temporarily disable flaky planner test

* Add NOLINTNEXTLINE

* Remove old sim config and add TODOs for trajectory utils
